font/src/directwrite/mod.rs

```diff
@@ -13,30 +13,168 @@
 };
 
+/// This struct contains the properties from dwrote::Font needed in the rest of this module.
+///
+/// We use this struct rather than dwrote::Font for performance reasons: the getter methods on
+/// dwrote::Font need to do things like String conversion from C++ widestrings to Rust's
+/// UTF-8 strings.
+pub struct Font {
+    face: FontFace,
+    family_name: String,
+    weight: FontWeight,
+    style: FontStyle,
+    stretch: FontStretch,
+}
+
 pub struct DirectWriteRasterizer {
-    fonts: Vec<dwrote::FontFace>,
+    fonts: HashMap<FontKey, Font>,
+    keys: HashMap<FontDesc, FontKey>,
     device_pixel_ratio: f32,
+    collection: FontCollection,
+    fallback: Option<FontFallback>,
+}
+
+impl DirectWriteRasterizer {
+    fn rasterize_glyph(
+        &self,
+        face: &FontFace,
+        size: Size,
+        c: char,
+    ) -> Result<RasterizedGlyph, Error> {
+        let glyph_index = self.get_glyph_index(face, c)?;
+
+        let em_size = em_size(size);
+
+        let glyph_run = DWRITE_GLYPH_RUN {
+            fontFace: unsafe { face.as_ptr() },
+            fontEmSize: em_size,
+            glyphCount: 1,
+            glyphIndices: &glyph_index,
+            glyphAdvances: &0.0,
+            glyphOffsets: &GlyphOffset::default(),
+            isSideways: 0,
+            bidiLevel: 0,
+        };
+
+        let rendering_mode = face.get_recommended_rendering_mode_default_params(
+            em_size,
+            self.device_pixel_ratio,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+        );
+
+        let glyph_analysis = GlyphRunAnalysis::create(
+            &glyph_run,
+            self.device_pixel_ratio,
+            None,
+            rendering_mode,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+            0.0,
+            0.0,
+        )
+        .map_err(Error::DirectWriteError)?;
+
+        let bounds = glyph_analysis
+            .get_alpha_texture_bounds(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1)
+            .map_err(Error::DirectWriteError)?;
+
+        let buf = glyph_analysis
+            .create_alpha_texture(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1, bounds)
+            .map_err(Error::DirectWriteError)?;
+
+        Ok(RasterizedGlyph {
+            c,
+            width: (bounds.right - bounds.left) as i32,
+            height: (bounds.bottom - bounds.top) as i32,
+            top: -bounds.top,
+            left: bounds.left,
+            buf: BitmapBuffer::RGB(buf),
+        })
+    }
+
+    fn get_loaded_font(&self, font_key: FontKey) -> Result<&Font, Error> {
+        self.fonts.get(&font_key).ok_or(Error::FontNotLoaded)
+    }
+
+    fn get_glyph_index(&self, face: &FontFace, c: char) -> Result<u16, Error> {
+        let idx = *face
+            .get_glyph_indices(&[c as u32])
+            .first()
+            // DirectWrite returns 0 if the glyph does not exist in the font
+            .filter(|glyph_index| **glyph_index != 0)
```

Comment:
Is there any way for this to return `0` as the first element of the indices, but then continue with a usable non-zero value later?

Reply:
As per the comment, which comes from the [remarks on GetGlyphIndices](https://docs.microsoft.com/en-gb/windows/win32/api/dwrite/nf-dwrite-idwritefontface-getglyphindices#remarks) `0` is the value returned when the glyph is not in the font, so that's definitely the error condition.

If you're concerned about the length - `get_glyph_indices()` will always return a `Vec` of length equal to the input slice - so this vec will always have only one element in it.

Perhaps you prefer this alternative implementation:

```
        match face.get_glyph_indices(&[c as u32]).as_slice() {
            [0] => Err(Error::MissingGlyph(c)),
            [i] => Ok(*i),

            // get_glyph_indices will always return a Vec with length equal to the input slice
            _ => unreachable!()
        }
```

Reply:
I'm fine with the implementation, I was just concerned if it might be necessary to check for later elements in the vec.